gum/backend-qnx/gummemory-qnx.c

```diff
@@ -0,0 +1,216 @@
+ *
+ * Licence: wxWindows Library Licence, Version 3.1
+ */
+
+#include "gummemory.h"
+#include "gumqnx-priv.h"
+
+#include <gio/gio.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/procfs.h>
+#include <stdio.h>
+
+GumPageProtection
+_gum_page_protection_from_posix (const gint flags)
+{
+  GumPageProtection prot = GUM_PAGE_NO_ACCESS;
+
+  if (flags & PROT_READ)
+    prot |= GUM_PAGE_READ;
+  if (flags & PROT_WRITE)
+    prot |= GUM_PAGE_WRITE;
+  if (flags & PROT_EXEC)
+    prot |= GUM_PAGE_EXECUTE;
+
+  return prot;
+}
+
+static gboolean
+gum_memory_get_protection (GumAddress address,
+                           gsize n,
+                           gsize * size,
+                           GumPageProtection * prot)
+{
+  gboolean success;
+  gint fd, res;
+  procfs_mapinfo * mapinfos;
+  gint num_mapinfos;
+  gpointer start, end;
+  gint i = 0;
+
+  if (size == NULL || prot == NULL)
+  {
+    gsize ignored_size;
+    GumPageProtection ignored_prot;
+
+    return gum_memory_get_protection (address, n,
+        (size != NULL) ? size : &ignored_size,
+        (prot != NULL) ? prot : &ignored_prot);
+  }
+
+  if (n > 1)
+  {
+    GumAddress page_size, start_page, end_page, cur_page;
+
+    page_size = gum_query_page_size ();
+
+    start_page = address & ~(page_size - 1);
+    end_page = (address + n - 1) & ~(page_size - 1);
+
+    success = gum_memory_get_protection (start_page, 1, NULL, prot);
+    if (success)
+    {
+      *size = page_size - (address - start_page);
+      for (cur_page = start_page + page_size;
+          cur_page != end_page + page_size;
+          cur_page += page_size)
+      {
+        GumPageProtection cur_prot;
+
+        if (gum_memory_get_protection (cur_page, 1, NULL, &cur_prot)
+            && (cur_prot != GUM_PAGE_NO_ACCESS || *prot == GUM_PAGE_NO_ACCESS))
+        {
+          *size += page_size;
+          *prot &= cur_prot;
+        }
+        else
+        {
+          break;
+        }
+      }
+      *size = MIN (*size, n);
+    }
+
+    return success;
+  }
+
+
+  success = FALSE;
+  *size = 0;
+  *prot = GUM_PAGE_NO_ACCESS;
+
+  fd = open ("/proc/self/as", O_RDONLY);
+  g_assert (fd != -1);
+
+  res = devctl (fd, DCMD_PROC_PAGEDATA, 0, 0, &num_mapinfos);
+  g_assert (res == 0);
+
+  mapinfos = g_malloc (sizeof (procfs_mapinfo) * num_mapinfos);
+
+  res = devctl (fd, DCMD_PROC_PAGEDATA, mapinfos,
+      sizeof (procfs_mapinfo) * num_mapinfos, &num_mapinfos);
+  g_assert (res == 0);
+
+  while (i != num_mapinfos)
+  {
+    start = GSIZE_TO_POINTER (mapinfos[i].vaddr & 0xffffffff);
+    end = start + mapinfos[i].size;
+
+    if (GUM_ADDRESS (start) > address)
+      break;
+    else if (address >= GUM_ADDRESS (start) &&
+        address + n -1 < GUM_ADDRESS (end))
+    {
+      success = TRUE;
+      *size = 1;
+
+      *prot = _gum_page_protection_from_posix (mapinfos[i].flags);
+      break;
+    }
+
+    i++;
+  }
+
+  close (fd);
+  g_free (mapinfos);
+  return success;
```

Comment:
I'd say add a newline before the return to separate these two logical statement groups. It would also be great to swap the order of the `close` and the `g_free`, so they're symmetrical.